rfc/system/0000-caching-of-offloaded-objects.md

```diff
@@ -0,0 +1,161 @@
+# [RFC] Caching of offloaded types
+
+**Authors:**
+
+- @eapolinario
+
+## 1 Executive Summary
+
+We propose a way to override the default behavior of [caching task executions](https://docs.flyte.org/projects/cookbook/en/latest/auto/core/flyte_basics/task_cache.html), enabling cache-by-value semantics for certain categories of objects.
+
+## 2 Motivation
+
+The behavior displayed by the cache, in some cases, does not match the users intuitions. 
+
+A good example of this involves pandas dataframes. As the system currently stands, we cannot honor the promise that cache keys will take into the account the value of the dataframes, because of how we represent dataframes in the backend.
+
+This RFC discusses how to extend this model to allow users to override the representation of objects used for caching purposes.
+
+## 3 Proposed Implementation
+
+We will expose a new field in [Literal](https://github.com/flyteorg/flyteidl/blob/master/protos/flyteidl/core/literals.proto#L68-L79) objects called `hash`, which will be used to represent that literal in cache key calculations. 
+
+Each client will then define the mechanics of setting that field when it's appropriate. Specifically for flytekit, we are going to use the extensions to the typing module proposed in https://www.python.org/dev/peps/pep-0593/, more specifically [`typing.Annotated`](https://docs.python.org/3/library/typing.html#typing.Annotated). 
+
+The following example illustrates how these annotated return objects are going to look like:
+
+
+```python
+@task
+def foo(a: int, b: str) -> Annotated[pd.DataFrame, HashMethod(hash_pandas_dataframe_function) :
+    df = pd.Dataframe(...)
+    ...
+    return df
+
+@task(cached=True, version="1.0")
+def bar(df: pd.Dataframe) -> int:
+    ...
+    
+@workflow
+def wf(a: int, b: str):
+    df = foo(a=a, b=b)
+    # Note that:
+    #   1. the return type of `foo` wraps around a pandas datataframe and adds some metadata to it.
+    #   2. the task `bar` is marked as cached and since the dataframe returned by `foo` overrides its hash
+    #      we will check the cache using the dataframe's hash as opposed to the literal representation.
+    v = bar(df=df) 
+```
+
+It's worth noting that this is a strictly opt-in feature, controlled at the level of Type Transformers. In other words, annotating types for which Type Transformers are not marked as opted in will be a no-op.
```

Comment:
> would it be a fair assumption, on users side, to say that "built-in" types are always by value?

That's true, but it happened almost by accident, i.e. we never exposed (nor enforced) the notion that values were being cached by value or by reference.

Also, we shouldn't be too prescriptive about hashing complex types for two reasons:
1. cost: calculating hashes might be expensive
2. No single way of calculating the hash. 
    i. We can certainly offer a few default implementations (and be upfront about the probability of collisions, etc)


Does this make sense? Can you clarify what we could gain if we made this assumption of all types will be cached by value the default?

Reply:
> Can you clarify what we could gain if we made this assumption of all types will be cached by value the default?

Mainly that from the feedback we heard a few times (and maybe it's overblown) that this was what the existing UX suggested to them... the reason I want to discuss this now is if we decide to offer it by default to some types, we will need an opt-out experience... and I think it might be worth writing down a few examples of each....

Reply:
This is the first time we're even making the distinction between the two caching semantics. I think it's fair to assume (and make it clear in the docs) that we use cache-by-value everywhere unless specified, in which case users will be able to opt-in to this experience of providing cache-by-value semantics for specific types.